File: xmlBlaster/src/c++/util/Log.cc
/*----------------------------------------------------------------------------
Name:      Log.cc
Project:   xmlBlaster.org
Copyright: xmlBlaster.org, see xmlBlaster-LICENSE file
Comment:   Handling the Client data
Version:   $Id: Log.cc,v 1.1 2000/07/06 22:55:44 laghi Exp $
----------------------------------------------------------------------------*/

#include <util/Log.h>

namespace util {

   Log::Log(int args, char *argc[]) {
      bool isNewProperty = false;
      if (numOfImplementations_ == 0) {
	 properties_ = new Property(args, argc);
	 isNewProperty = true;
      }
      else {
	 if ((args != 0) && (argc != 0)) {
  	    delete properties_;
  	    properties_ = new Property(args, argc);
	    isNewProperty = true;
	 }
      }

      if (isNewProperty) { // all static stuff here ...
	 CALLS   = true;
	 TIME    = true;
	 TRACE   = true;
	 DUMP    = true;
	 currentLogFormat       = "{0} {1} {2}: {3}";
	 logFormatPropertyRead  = false;
	 logLevel_ = L_PANIC | L_ERROR | L_WARN | L_INFO;
	 string path = properties_->getProperty("HOME");
	 string name = "xmlBlaster.properties";
	 // introduce exceptions here if needed...
	 int ret = properties_->loadPropsFromFile(name, path); 
	 if ( ret == -1) {
	    path = properties_->getProperty("XMLBLASTER_HOME");
	    ret  = properties_->loadPropsFromFile(name,path);
	    if (ret == -1) 
	       error("reading properties: ", "could not find the file");
	 }
      }
      // the following stuff does not matter if reinitialized many times
      ME      = "Log";
      timeX   = "TIME ";
      callsX  = "CALL ";
      traceX  = "TRACE";
      plainX  = "     ";
      infoX   = "INFO ";
      warnX   = "WARN ";
      errorX  = "ERROR";
      panicX  = "PANIC";
      exitX   = "EXIT ";
#ifdef _TERM_WITH_COLORS_
      timeE   = string(LTGREEN_BLACK) + "TIME " + ESC;
      callsE  = string(BLACK_LTGREEN) + "CALL " + ESC;
      traceE  = string(WHITE_BLACK  ) + "TRACE" + ESC;
      plainE  = string(WHITE_BLACK  ) + "     " + ESC;
      infoE   = string(GREEN_BLACK  ) + "INFO " + ESC;
      warnE   = string(YELLOW_BLACK ) + "WARN " + ESC;
      errorE  = string(RED_BLACK    ) + "ERROR" + ESC;
      panicE  = string(BLACK_RED    ) + "PANIC" + ESC;
      exitE   = string(GREEN_BLACK  ) + "EXIT " + ESC;
#endif // _TERM_WITH_COLORS_      
      numOfImplementations_++;
   }

   
   Log::~Log() {
      numOfImplementations_--;
      if (numOfImplementations_ == 0) {
	 if (properties_) delete properties_;
	 properties_ = 0;
      }
   }
   
   void Log::setDefaultLogLevel() {
      logLevel_ = L_PANIC | L_ERROR | L_WARN | L_INFO;
      setPreLogLevelCheck();
   }

   
   void Log::setLogLevel(int level) {
      logLevel_ = level;
      setPreLogLevelCheck();
   }
   
   void Log::setLogLevel(int argc, char *args[]) {
      if ((properties_->findArgument(argc, args, "-?") > 0) || 
	  (properties_->findArgument(argc, args, "-h") > 0)) {
	 usage();
	 return;
      }
      initialize();
   }


   void Log::removeLogLevel(string logLevel) {
      int level = logLevelToBit(logLevel);
      logLevel_ = (logLevel_ & ~level);
      setPreLogLevelCheck();
   }
   
   
   void Log::addLogLevel(string logLevel) {
      int level = logLevelToBit(logLevel);
      logLevel_ = (logLevel_ | level);
      setPreLogLevelCheck();
   }
            

   void Log::setPreLogLevelCheck() {
      CALLS = TIME = TRACE = DUMP = false;
      if (logLevel_ & L_CALLS) CALLS = true;
      if (logLevel_ & L_TIME ) TIME  = true;
      if (logLevel_ & L_TRACE) TRACE = true;
      if (logLevel_ & L_DUMP ) DUMP  = true;
   }
      
   
   string Log::bitToLogLevel(int level) {
      string sb = "";
      if (level & L_PANIC) sb += "PANIC";
      if (level & L_ERROR) sb += " | ERROR";
      if (level & L_WARN ) sb += " | WARN";
      if (level & L_INFO ) sb += " | INFO";
      if (level & L_CALLS) sb += " | CALLS";
      if (level & L_TIME ) sb += " | TIME";
      if (level & L_TRACE) sb += " | TRACE";
      if (level & L_DUMP ) sb += " | DUMP";
      return sb;
   }
   

   void Log::panic(const string &instance, const string &text) {
      if (logLevel_ & L_PANIC) {
#ifdef _TERM_WITH_COLORS_
	 log(panicE, L_PANIC, instance, text);
#else
	 log(panicX, L_PANIC, instance, text);
#endif
	 cerr << text << endl;
	 numErrorInvocations++;
	 // displayStatistics();
	 exitLow(1);
      }
   }


   void Log::exit(const string &instance, const string &text) {
#ifdef _TERM_WITH_COLORS_
      log(exitE, L_EXIT, instance, text);
#else
      log(exitX, L_EXIT, instance, text);
#endif
      displayStatistics();
      exitLow(0);
   }
   

   void Log::info(const string &instance, const string &text) {
      if (logLevel_ & L_INFO) {
#ifdef _TERM_WITH_COLORS_
	 log(infoE, L_INFO, instance, text);
#else
	 log(infoX, L_INFO, instance, text);
#endif
      }
   }
      

   void Log::warning(const string &instance, const string &text) {
      if(logLevel_ & L_WARN) {
	 numWarnInvocations++;
#ifdef _TERM_WITH_COLORS_
	 log(warnE, L_WARN, instance, text);
#else
	 log(warnX, L_WARN, instance, text);
#endif
      }
   }
   
      
   void Log::error(const string &instance, const string &text) {
      if(logLevel_ & L_PANIC) {
	 numErrorInvocations++;
#ifdef _TERM_WITH_COLORS_
	 log(errorE, L_ERROR, instance, text);
#else
	 log(errorX, L_ERROR, instance, text);
#endif
      }
   }

   
   void Log::plain(const string &instance, const string &text) {
      log("", L_PLAIN, "", text);
   }

   
   void Log::dump(const string &instance, const string &text) {
      if((logLevel_ & L_DUMP) != 0) {
	 log("", L_DUMP, instance, text);
//	    log("", L_DUMP, instance, Memory.getStatistic());
      }
   }
   

   void Log::trace(const string &instance, const string &text) {
      if(logLevel_ & L_TRACE) {
#ifdef _TERM_WITH_COLORS_
	 log(traceE, L_TRACE, instance, text);
#else
	 log(traceX, L_TRACE, instance, text);
#endif
      }
   }

   
   void Log::calls(const string &instance, const string &text) {
      if(logLevel_ & L_CALLS) {
#ifdef _TERM_WITH_COLORS_
	 log(callsE, L_CALLS, instance, text);
#else
	 log(callsX, L_CALLS, instance, text);
#endif
      }
   }
      
      
   void Log::time(const string &instance, const string &text) {
	 if(logLevel_ & L_TIME) {
#ifdef _TERM_WITH_COLORS_
	    log(timeE, L_TIME, instance, text);
#else
	    log(timeX, L_TIME, instance, text);
#endif
	 }
      }
   
   
   void Log::usage() {
      plain(ME, "");
      plain(ME, "Logging options:");
      plain(ME, "   +trace              Show code trace.");
      plain(ME, "   +dump               Dump internal state.");
      plain(ME, "   +calls              Show important method entries");
      plain(ME, "   +time               Display some performance data.");
      plain(ME, "   -logFile <fileName> Log to given file instead to console.");
      plain(ME, "");
   }
   
   
   void Log::displayStatistics() {
      char       buffer[512];
      ostrstream out(buffer, 511);
//	 Log.info(ME, Memory.getStatistic());
#ifdef _TERM_WITH_COLORS_
      if (numErrorInvocations>0) {
	 out << BLACK_RED << "There were " << numErrorInvocations;
	 out << " ERRORS and " << numWarnInvocations << " WARNINGS";
	 out << ESC << (char)0;
      }
      else if (numWarnInvocations>0) {
	 out << BLACK_PINK << "There were " << numErrorInvocations;
	 out << " ERRORS and " << numWarnInvocations << " WARNINGS";
	 out << ESC << (char)0;
      }
      else {
	 out << BLACK_GREEN << "No errors/warnings were reported";
	 out << ESC << (char)0;
      }
#else
      if (numErrorInvocations>0 || numWarnInvocations>0) {
	 out << "There were " << numErrorInvocations << " ERRORS and ";
	 out << numWarnInvocations << " WARNINGS" << (char)0;
      }
      else
	 out << "No errors/warnings were reported" << (char)0;
#endif
      info(ME, buffer);
   }
   
   
   void Log::printStack() {
      cerr << "sorry, no Stack aviable" << endl;
   }
   

   void Log::initialize() {
      logFormatPropertyRead = true;
      // Given flag -info switches off Log.info messages:
      if (properties_->propertyExists("info")) {
	 if (properties_->getBoolProperty("info", false)) 
	    addLogLevel("INFO");
	 else removeLogLevel("INFO");
      }
      if (properties_->propertyExists("warn")) {
	 if (properties_->getBoolProperty("warn", false)) 
	    addLogLevel("WARN");
	 else removeLogLevel("WARN");
      }
      if (properties_->propertyExists("error")) {
	 if (properties_->getBoolProperty("error", false)) 
	    addLogLevel("ERROR");
	 else removeLogLevel("ERROR");
      }
      if (properties_->propertyExists("calls")) {
	 if (properties_->getBoolProperty("calls", false)) 
	    addLogLevel("CALLS");
	 else removeLogLevel("CALLS");
      }
      if (properties_->propertyExists("time")) {
	 if (properties_->getBoolProperty("time", false)) 
	    addLogLevel("TIME");
	 else removeLogLevel("TIME");
      }
      if (properties_->propertyExists("trace")) {
	 if (properties_->getBoolProperty("trace", false)) 
	    addLogLevel("TRACE");
	 else removeLogLevel("TRACE");
      }
      if (properties_->propertyExists("dump")) {
	 if (properties_->getBoolProperty("dump", false)) 
	    addLogLevel("DUMP");
	 else removeLogLevel("DUMP");
      }
      
      if (properties_->getBoolProperty("+calls", false)) 
	 addLogLevel("CALLS");
      if (properties_->getBoolProperty("+time", false)) 
	 addLogLevel("TIME");
      if (properties_->getBoolProperty("+trace", false)) 
	 addLogLevel("TRACE");
      if (properties_->getBoolProperty("+dump", false))
	 addLogLevel("DUMP");
      
      // format: {0}:{1}:{2}:{3}    <timestamp>:<levelStr>:<instance>:<text>
      currentLogFormat = properties_->getStringProperty("LogFormat", 
						       currentLogFormat);
      string tmp = properties_->getStringProperty("LogFormat.Date","MEDIUM");
//	 if (tmp == "SHORT") lookAndFeelDate = java.text.DateFormat.SHORT;
//  	 else if (tmp.equals("MEDIUM")) 
//  	    lookAndFeelDate = java.text.DateFormat.MEDIUM;
//  	 else if (tmp.equals("LONG")) 
//  	    lookAndFeelDate = java.text.DateFormat.LONG;
//  	 else if (tmp.equals("FULL")) 
//  	    lookAndFeelDate = java.text.DateFormat.FULL;
//  	 tmp = properties_.getStringProperty("LogFormat.Time","MEDIUM");
//  	 if (tmp.equals("SHORT")) 
//  	    lookAndFeelTime = java.text.DateFormat.SHORT;
//  	 else if (tmp.equals("MEDIUM")) 
//  	    lookAndFeelTime = java.text.DateFormat.MEDIUM;
//  	 else if (tmp.equals("LONG")) 
//  	    lookAndFeelTime = java.text.DateFormat.LONG;
//  	 else if (tmp.equals("FULL")) 
//  	    lookAndFeelTime = java.text.DateFormat.FULL;
      
      string la = properties_->getStringProperty("LogFormat.Language","");
      string co = properties_->getStringProperty("LogFormat.Country", "");
//	 if (la != null && co != null) country = new Locale(la, co);
      
//  	 String fileName = properties_.getProperty("logFile", (String)null);
//  	 if (fileName != null)
//  	    Log.logToFile(fileName);
   }
   
   
   string Log::getTime() {
      // adapt it here to the correct time format (locales) ?!
      time_t theTime;
      std::time(&theTime);
      string timeStr = std::ctime(&theTime), ret;
      // eliminate new lines (if any)
      int pos = timeStr.find("\n");
      if (pos < 0) return timeStr;
      ret.assign(timeStr, 0, pos);
      return ret;
   }

   Property *Log::properties_        = 0;
   int    Log::numWarnInvocations    = 0;
   int    Log::numErrorInvocations   = 0;
   int    Log::numOfImplementations_ = 0;
   int    Log::logLevel_             = 0;
   string Log::currentLogFormat      = "";
   bool   Log::logFormatPropertyRead = false;
   bool   Log::CALLS = false;
   bool   Log::TIME  = false;    
   bool   Log::TRACE = false;
   bool   Log::DUMP  = false;


}; // end of namespace

File: xmlBlaster/src/java/org/xmlBlaster/util/Log.java
/*------------------------------------------------------------------------------
Name:      RequestBroker.java
Project:   xmlBlaster.org
Copyright: xmlBlaster.org, see xmlBlaster-LICENSE file
Comment:   Handling the Client data
Version:   $Id: Log.java,v 1.20 1999/12/09 13:28:37 ruff Exp $
------------------------------------------------------------------------------*/
package org.xmlBlaster.util;


/**
 * Logging output
 */
public class Log
{
   private final static String ME = "Log";

   /**
    * Produce logging output on important method calls. 
    * <p />
    * Switch CALLS mode to <code>false</code> for performance reasons<br />
    * or switch it to <code>true</code> for debugging reasons. 
    * <p />
    * Setting to false:<br />
    * <code>if (Log.CALLS) Log.trace(....); </code> -> dead code elimination
    * <p />
    * The same applies for TIME, TRACE and DUMP
    */
   public final static boolean CALLS = true;  // trace method calls
   /**
    * Performance logging output true/false
    */
   public final static boolean TIME  = true;  // trace performance
   /**
    * Fine grained code logging output true/false
    */
   public final static boolean TRACE = true;  // trace application flow
   /**
    * Dump internal xmlBlaster state in xml format, true/false
    */
   public final static boolean DUMP  = true;  // dump internal state

   /**
    * Logging levels
    */
   private static final int L_NOLOG = 0;
   private static final int L_PANIC = 10;  // Abbruch der Aktion
   private static final int L_ERROR = 20;  // Fehler ohne Abbruch
   private static final int L_WARN  = 30;  // Warung
   private static final int L_INFO  = 40;  // reine Information
   private static final int L_DEBUG = 50;  // für Werte zum Testen
   private static final int L_TRACE = 60;  // ganz feiner Log

   private static int LOGLEVEL = 60;

   /**
    * Adjust here your local look and feel
    */
   private final static int lookAndFeelDate = java.text.DateFormat.MEDIUM;
   private final static int lookAndFeelTime = java.text.DateFormat.MEDIUM;
   private final static java.util.Locale country = java.util.Locale.US;

   /**
    * Counter for occurred warnings/errors
    */
   private static int numWarnInvocations = 0;
   private static int numErrorInvocations = 0;

   /**
    * colors foreground/background
    */
   private final static String ESC          = "\033[0m";     // Reset color to original values
   private final static String BOLD         = "\033[1m";

   private final static String RED_BLACK    = "\033[31;40m";
   private final static String GREEN_BLACK  = "\033[32;40m";
   private final static String YELLOW_BLACK = "\033[33;40m";
   private final static String BLUE_BLACK   = "\033[34;40m";
   private final static String PINK_BLACK   = "\033[35;40m";
   private final static String LTGREEN_BLACK= "\033[36;40m";
   private final static String WHITE_BLACK  = "\033[37;40m";

   private final static String WHITE_RED    = "\033[37;41m";
   private final static String BLACK_RED    = "\033[30;41m";
   private final static String BLACK_GREEN  = "\033[40;42m";
   private final static String BLACK_PINK   = "\033[40;45m";
   private final static String BLACK_LTGREEN= "\033[40;46m";


   /**
    * Output text for different logging levels
    */
   private final static StringBuffer INSTANCE_SEPERATOR = new StringBuffer(":  ");
   private final static StringBuffer timeX  = new StringBuffer("TIME  ");
   private final static StringBuffer callsX = new StringBuffer("CALL  ");
   private final static StringBuffer traceX = new StringBuffer("TRACE ");
   private final static StringBuffer plainX = new StringBuffer("      ");
   private final static StringBuffer infoX  = new StringBuffer("INFO  ");
   private final static StringBuffer warnX  = new StringBuffer("WARN  ");
   private final static StringBuffer errorX = new StringBuffer("ERROR ");
   private final static StringBuffer panicX = new StringBuffer("PANIC ");
   private final static StringBuffer exitX  = new StringBuffer("EXIT  ");

   /**
    * Colored output to xterm
    */
   private       static boolean withXtermEscapeColor = false;
   private final static StringBuffer INSTANCE_SEPERATOR_E = new StringBuffer(BOLD + ":" + ESC + "  "); // bold

   private final static StringBuffer timeE  = new StringBuffer(LTGREEN_BLACK+ "TIME " + ESC + ": ");
   private final static StringBuffer callsE = new StringBuffer(BLACK_LTGREEN+ "CALL " + ESC + ": ");
   private final static StringBuffer traceE = new StringBuffer(WHITE_BLACK  + "TRACE" + ESC + ": ");
   private final static StringBuffer plainE = new StringBuffer(WHITE_BLACK  + "     " + ESC + ": ");
   private final static StringBuffer infoE  = new StringBuffer(GREEN_BLACK  + "INFO " + ESC + ": ");
   private final static StringBuffer warnE  = new StringBuffer(YELLOW_BLACK + "WARN " + ESC + ": ");
   private final static StringBuffer errorE = new StringBuffer(RED_BLACK    + "ERROR" + ESC + ": ");
   private final static StringBuffer panicE = new StringBuffer(BLACK_RED    + "PANIC" + ESC + ": ");
   private final static StringBuffer exitE  = new StringBuffer(GREEN_BLACK  + "EXIT " + ESC + ": ");


   static {
      String osName = System.getProperty("os.name");     // "Linux" "Windows NT" ...
      if (osName.startsWith("Window"))
         withXtermEscapeColor = false;
      else
         withXtermEscapeColor = true;

   }

   /**
    * Sets the loglevel
    */
   public final static void setLogLevel(int LEVEL)
   {
    LOGLEVEL = LEVEL;
   }

  /**
   * Gets the loglevel 0,10,20,30,40,50,60
   */
  public final static int getLogLevel()
  {
    return LOGLEVEL;
  }

   /**
    * Create the Date/Time informations
    *
    * @return example: 1999-11-01 10:41:03 INFO
    */
   public final static StringBuffer logHeader(StringBuffer levelStr)
   {
      if (levelStr == null) return new StringBuffer("");

      java.util.Date currentDate = new java.util.Date();

      java.text.DateFormat df = java.text.DateFormat.getDateTimeInstance(lookAndFeelDate, lookAndFeelTime, country);

      String timeStr = df.format(currentDate);

      StringBuffer time = new StringBuffer(timeStr.length() + 4 + levelStr.length());

      return time.append(timeStr).append(" ").append(levelStr);
   }


   /**
    * Log example:
    *
    *   1999-11-01 10:41:03 INFO : xmlBlaster is ready to blast
    */
   public final static void log(final StringBuffer levelStr, String instance, String text)
   {
      // System.err.println(logHeader(levelStr).append(instance).append(withXtermEscapeColor ? INSTANCE_SEPERATOR_E : INSTANCE_SEPERATOR).append(text));
      System.err.println(logHeader(levelStr).append(" [").append(instance).append("]  ").append(text));
   }


   /**
    * The only way to stop the server
    * @param val exit code for operating system
    */
   private final static void exitLow(int val)
   {
      //Runtime.getRuntime().runFinalizersOnExit(true);
      Runtime.getRuntime().runFinalization();
      System.exit(val);
   }


   /**
    * Use this exit for errors
    */
   public final static void panic(String instance, String text)
   {
      if(LOGLEVEL >= L_PANIC)
      {
         log((withXtermEscapeColor) ? panicE : panicX, instance, text);
         numErrorInvocations++;
         displayStatistics();
         exitLow(1);
      }
   }


   /**
    * Exit without errors
    */
   public final static void exit(String instance, String text)
   {
      log((withXtermEscapeColor) ? exitE : exitX, instance, text);
      displayStatistics();
      exitLow(0);
   }


   /**
    * Use this for normal logging output
    */
   public final static void info(String instance, String text)
   {
      if(LOGLEVEL >= L_INFO)
      {
         log((withXtermEscapeColor) ? infoE : infoX, instance, text);
      }
   }


   /**
    * Use this for logging output where the xmlBlaster administrator shall be informed<br />
    * for example a login denied event
    */
   public final static void warning(String instance, String text)
   {
      if(LOGLEVEL >= L_WARN)
      {
         numWarnInvocations++;
         log((withXtermEscapeColor) ? warnE : warnX, instance, text);
      }
   }
   /*
   public final static void warningThrow(String instance, String text) throws org.xmlBlaster.XmlBlasterException
   {
      warning(instance, text);
      throw new org.xmlBlaster.XmlBlasterException(instance, text);
   }
   */

   /**
    * Use this for internal xmlBlaster errors reporting
    */
   public final static void error(String instance, String text)
   {
      if(LOGLEVEL >= L_ERROR)
      {
         numErrorInvocations++;
         log((withXtermEscapeColor) ? errorE : errorX, instance, text);
      }
   }
   /*
   public final static void errorThrow(String instance, String text) throws org.xmlBlaster.XmlBlasterException
   {
      error(instance, text);
      throw new org.xmlBlaster.XmlBlasterException(instance, text);
   }
   */

   /*
    * Log without time/date
    */
   public final static void plain(String instance, String text)
   {
      log(null, instance, text);
   }

   /*
    * Log without time/date
    */
   public final static void dump(String instance, String text)
   {
      log(null, instance, text);
   }

   /**
    * Tracing execution
    */
   public final static void trace(String instance, String text)
   {
      if(LOGLEVEL >= L_TRACE)
      {
         log((withXtermEscapeColor) ? traceE : traceX, instance, text);
      }
   }

   /**
    * Tracing when entering methods
    */
   public final static void calls(String instance, String text)
   {
      log((withXtermEscapeColor) ? callsE : callsX, instance, text);
   }

   /**
    * Output of performant measurements (elapsed milliseconds)
    */
   public final static void time(String instance, String text)
   {
      log((withXtermEscapeColor) ? timeE : timeX, instance, text);
   }

   /**
    * Display some statistic on exit
    */
   public final static void displayStatistics()
   {
      if (withXtermEscapeColor) {
         if (numErrorInvocations>0)
            Log.info(ME, BLACK_RED + "There were " + numErrorInvocations + " ERRORS and " + numWarnInvocations + " WARNINGS" + ESC);
         else if (numWarnInvocations>0)
            Log.info(ME, BLACK_PINK + "There were " + numErrorInvocations + " ERRORS and " + numWarnInvocations + " WARNINGS" + ESC);
         else
            Log.info(ME, BLACK_GREEN + "No errors/warnings were reported" + ESC);
      }
      else {
         if (numErrorInvocations>0 || numWarnInvocations>0)
            Log.info(ME, "There were " + numErrorInvocations + " ERRORS and " + numWarnInvocations + " WARNINGS");
         else
            Log.info(ME, "No errors/warnings were reported");
      }
   }




   /**
    * Only for testing
    *    java org.xmlBlaster.util.Log
    */
   public static void main(String args[]) throws Exception
   {
      String me = "Log-Tester";
                        System.out.println("LOGLEVEL : "+getLogLevel());
      setLogLevel(20);
                        System.out.println("LOGLEVEL : "+getLogLevel());
         Log.panic(me, "Panic is not shown");
         Log.error(me, "Error ...");
      setLogLevel(60);
      Log.calls(me, "Entering method main()");
      Log.time(me, "Elapsed time = 34 milli seconds");
      Log.trace(me, "TRACE: Entering method xy");
      Log.info(me, "Successfully loaded data");
      Log.warning(me, "your input is strange");
      Log.error(me, "given User unknown");
      Log.error(me, "another error");
      Log.panic(me, "Unrecoverable Error - good bye");
   }

}


